bottlecap/src/traces/propagation/text_map_propagator.rs

```diff
@@ -0,0 +1,508 @@
+            return None;
+        }
+
+        trace_id.parse::<u64>().ok()
+    }
+
+    fn extract_parent_id(carrier: &dyn Extractor) -> Option<u64> {
+        let parent_id = carrier.get(DATADOG_PARENT_ID_KEY)?;
+
+        parent_id.parse::<u64>().ok()
+    }
+
+    fn extract_sampling_priority(carrier: &dyn Extractor) -> Option<i8> {
+        let sampling_priority = carrier.get(DATADOG_SAMPLING_PRIORITY_KEY)?;
+
+        sampling_priority.parse::<i8>().ok()
+    }
+
+    fn extract_origin(carrier: &dyn Extractor) -> Option<String> {
+        let origin = carrier.get(DATADOG_ORIGIN_KEY)?;
+        Some(origin.to_string())
+    }
+
+    fn extract_tags(carrier: &dyn Extractor) -> HashMap<String, String> {
+        let carrier_tags = carrier.get(DATADOG_TAGS_KEY).unwrap_or_default();
+        let mut tags: HashMap<String, String> = HashMap::new();
+
+        // todo:
+        // - trace propagation disabled
+        // - trace propagation max lenght
+
+        let pairs = carrier_tags.split(',');
+        for pair in pairs {
+            if let Some((k, v)) = pair.split_once('=') {
+                // todo: reject key on tags extract reject
+                if k.starts_with("_dd.p") {
+                    tags.insert(k.to_string(), v.to_string());
+                }
+            }
+        }
+
+        // Handle 128bit trace ID
+        if !tags.is_empty() {
+            if let Some(trace_id_higher_order_bits) =
+                carrier.get(DATADOG_HIGHER_ORDER_TRACE_ID_BITS_KEY)
+            {
+                if !Self::higher_order_bits_valid(trace_id_higher_order_bits) {
+                    warn!("Malformed Trace ID: {trace_id_higher_order_bits} Failed to decode trace ID from carrier.");
+                    tags.insert(
+                        DATADOG_PROPAGATION_ERROR_KEY.to_string(),
+                        format!("malformed tid {trace_id_higher_order_bits}"),
+                    );
+                    tags.remove(DATADOG_HIGHER_ORDER_TRACE_ID_BITS_KEY);
+                }
+            }
+        }
+
+        if !tags.contains_key(DATADOG_SAMPLING_DECISION_KEY) {
+            tags.insert(DATADOG_SAMPLING_DECISION_KEY.to_string(), "-3".to_string());
+        }
+
+        tags
+    }
+
+    fn higher_order_bits_valid(trace_id_higher_order_bits: &str) -> bool {
+        if trace_id_higher_order_bits.len() != 16 {
+            return false;
+        }
+
+        match u64::from_str_radix(trace_id_higher_order_bits, 16) {
+            Ok(_) => {}
+            Err(_) => return false,
+        }
+
+        true
+    }
+}
+
+struct Traceparent {
+    sampling_priority: i8,
+    trace_id: u128,
+    span_id: u64,
+}
+
+struct Tracestate {
+    sampling_priority: Option<i8>,
+    origin: Option<String>,
+    lower_order_trace_id: Option<String>,
+}
+
+#[derive(Clone, Copy)]
+pub struct TraceparentPropagator;
+
+impl TextMapPropagator for TraceparentPropagator {
+    fn extract(&self, carrier: &dyn Extractor) -> SpanContext {
```

Comment:
do we really need struct + trait + methods here for traceparent and tracestate propagators?
Can we get away with extract_parent and extract_state?

Reply:
> `traceparent and tracestate propagators`
Do you mean The `Traceparent` propagator which includse the extraction of the `tracestate` + `traceparent`?

If so, I tried using this `struct + trait + methods` in order to make the code easier to maintain, and also following the current OTel implemenetation, which I think was simple enough to understand, what do you think?